design/tlog-spilling.md.html

```diff
@@ -0,0 +1,590 @@
+in aggregate.  Spill-by-reference significantly reduces the volume of data
+written to the B-tree, and the less data that we write, the more we can batch
+versions to be written together.
+
+************************************************************************
+*                              DiskQueue                               *
+*                                                                      *
+*    -------- Index on disk --------     ---- Index in memory ----     *
+*   /                               \   /                         \    *
+* +-----------------------------------+-----------------------------+  *
+* |            Spilled Data           |       Most Recent Data      |  *
+* +-----------------------------------+-----------------------------+  *
+* lowest version                                      highest version  *
+*                                                                      *
+************************************************************************
+
+Spill-by-reference works by taking a larger range of versions, and building a
+single key-value pair per tag that describes where in the disk queue is every
+relevant commit for that tag.  Concretely, this takes the form
+`(tag, last_version) -> [(version, start, end, mutation bytes)]`, where...
+
+ * `tag` is the small integer representing the storage server this mutation batch is destined for.
+ * `last_version` is the last/maximum version contained in the value's batch.
+ * `version` is the version of the commit that this index entry points to.
+ * `start` is an index into the disk queue of where to find the beginning of the commit.
+ * `end` is an index into the disk queue of where the end of the commit is.
+ * `mutation_bytes` is the number of bytes in the commit that are relevant for this tag.
+
+And then writing only once per tag spilled into the B-tree for each iteration
+through spilling.
+
+Note that each tuple in the list represents a commit, and not a mutation.  This
+means that peeking spilled commits will involve reading mutations unrelated to
+the requested tag.  Alternatively, one could have each tuple represent a
+mutation within a commit, to prevent over-reading when peeking.  There exist
+pathological workloads for each strategy.  The purpose of this work is most
+importantly to support spilling of log router tags.  These exist on every
+mutation, so that it will get copied to other datacenters.  This is the exact
+pathological workload for recording each mutation individually, because it only
+increases the number of IO operations used to read the same amount of data.
+For a wider set of workloads, there's room to establish a heuristic as to when
+to record mutation(s) versus the entire commit, but performance testing hasn't
+surfaced this as important enough to include in the initial version of this
+work.
+
+Peeking now works by issuing a range read to the B-tree from `(tag, peek_begin)`
```

Comment:
This is for peeking spilled data right? I.E in the case where a storage node is not dead and is just trying to peek the most recent data it needs it will go directly to the in memory queues not the btree?